README.md
# network-manager

> Rust NetworkManager bindings

[![Current Release](https://img.shields.io/github/release/balena-io-modules/network-manager.svg?style=flat-square)](https://github.com/balena-io-modules/network-manager/releases/latest)
[![License](https://img.shields.io/github/license/balena-io-modules/network-manager.svg?style=flat-square)](https://github.com/balena-io-modules/network-manager/blob/master/LICENSE)
[![Issues](https://img.shields.io/github/issues/balena-io-modules/network-manager.svg?style=flat-square)](https://github.com/balena-io-modules/network-manager/issues)

<div align="center">
  <sub>an open source :satellite: project by <a href="https://balena.io">balena.io</a></sub>
</div>

---

## Support

If you're having any problem, please [raise an issue](https://github.com/balena-io-modules/network-manager/issues/new) on GitHub or [contact us](https://balena.io/community/), and the balena.io team will be happy to help.

---

## License

network-manager is free software, and may be redistributed under the terms specified in
the [license](https://github.com/balena-io-modules/network-manager/blob/master/LICENSE).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.13.3
## (2023-02-16)

* Remove unused VERSION file [Zahari Petkov]

# v0.13.2
## (2023-02-16)

* Switch repo type to `rust-module` to enable auto cargo versioning [Zahari Petkov]

# v0.13.1
## (2023-02-16)

* Add new AP security flags from NetworkManager [Zahari Petkov]

# v0.13.0
## (2023-02-16)

* Replace resin references with balena [maggie0002]

# v0.12.0
## (2023-02-16)

* Add missing NMDeviceTypes [maggie0002]

# v0.11.3
## (2023-02-15)

* Use Flowzone as a CI [Zahari Petkov]

## v0.11.2 - 2018-07-05

* Examples: Replace env! with crate_version! #122 [Robert Vojta]

## v0.11.1 - 2018-07-04

* Examples: CLI args parsing with clap crate #120 [Robert Vojta]

## v0.11.0 - 2018-06-28

* Wifi: WPA-Enterprise support #119 [majorz]

## v0.10.0 - 2018-03-14

* Expose `RequestScan` through `WiFiDevice::request_scan` #117 [majorz]

## v0.9.1 - 2018-03-06

* NetworkManager::get_state uses incorrect D-Bus type #114 [majorz]

## v0.9.0 - 2018-02-15

* Error handling with error-chain #110 [majorz]

## v0.8.6 - 2018-01-10

* Incomplete NM80211ApFlags enumeration definitions #108 [majorz]

## v0.8.5 - 2018-01-10

* Debug log level for expired D-Bus objects #104 [majorz]

## v0.8.4 - 2018-01-10

* Update maintainers list #102 [majorz]

## v0.8.3 - 2018-01-10

* Transitioning to CircleCI and running latest clippy and rustfmt #107 [majorz]

## v0.8.2 - 2017-10-20

* Do not include clippy as an optional dependency #100 [majorz]

## v0.8.1 - 2017-10-20

* Cargo keyword containing space #98 [majorz]

## v0.8.0 - 2017-10-20

* Rename project to `network-manager` #96 [majorz]

## v0.7.2 - 2017-10-20

* Fix clippy warnings about cloned Rc #94 [majorz]

## v0.7.1 - 2017-08-08

* Lower log level on get property of expired D-Bus object path #92 [majorz]

## v0.7.0 - 2017-08-07

* Expose Wi-Fi connection mode in ConnectionSettings [majorz]

## v0.6.3 - 2017-08-02

* Fix clippy warnings about long literals [majorz]

## v0.6.2 - 2017-07-04

* NetworkManager::get_connectivity fixed and added test [majorz]

## v0.6.1 - 2017-07-04

* Export publicly Connectivity and ServiceState [majorz]

## v0.6.0 - 2017-06-26

* Captive portal hotspot creation support [majorz]

## v0.5.0 - 2017-06-15

* Expose connection type in ConnectionSettings [majorz]

## v0.4.0 - 2017-06-15

* ASCII oriented SSID Debug formatting [majorz]

## v0.3.3 - 2017-06-15

* Stripping patch version from dependencies and bitflags catch up [majorz]

## v0.3.2 - 2017-06-15

* Keeping binary crates updated in Travis [majorz]

## v0.3.1 - 2017-06-15

* VersionBot config update for automatic Cargo.toml version bump [majorz]

## v0.3.0 - 2017-06-01

* Export publicly ConnectionState and DeviceState [majorz]

## v0.2.4 - 2017-06-01

* Clean up clippy warnings [majorz]

## v0.2.3 - 2017-05-31

* Treat clippy warnings as errors [majorz]

## v0.2.2 - 2017-05-30

* Logging in critical places of the library [majorz]

## v0.2.1 - 2017-05-30

* Custom rustfmt.toml settings for better auto-formatting [majorz]

## v0.2.0 - 2017-05-23

* New public API (spec in #58) [majorz]

## v0.1.0 2017-04-26

* Integrate versionbot


src/device.rs
use std::fmt;
use std::rc::Rc;

use dbus_nm::DBusNetworkManager;
use errors::*;

use wifi::{new_wifi_device, WiFiDevice};

#[derive(Clone)]
pub struct Device {
    dbus_manager: Rc<DBusNetworkManager>,
    path: String,
    interface: String,
    device_type: DeviceType,
}

impl Device {
    fn init(dbus_manager: &Rc<DBusNetworkManager>, path: &str) -> Result<Self> {
        let interface = dbus_manager.get_device_interface(path)?;

        let device_type = dbus_manager.get_device_type(path)?;

        Ok(Device {
            dbus_manager: Rc::clone(dbus_manager),
            path: path.to_string(),
            interface,
            device_type,
        })
    }

    pub fn device_type(&self) -> &DeviceType {
        &self.device_type
    }

    pub fn interface(&self) -> &str {
        &self.interface
    }

    pub fn get_state(&self) -> Result<DeviceState> {
        self.dbus_manager.get_device_state(&self.path)
    }

    pub fn as_wifi_device(&self) -> Option<WiFiDevice> {
        if self.device_type == DeviceType::WiFi {
            Some(new_wifi_device(&self.dbus_manager, self))
        } else {
            None
        }
    }

    /// Connects a Network Manager device.
    pub fn connect(&self) -> Result<DeviceState> {
        let state = self.get_state()?;

        match state {
            DeviceState::Activated => Ok(DeviceState::Activated),
            _ => {
                self.dbus_manager.connect_device(&self.path)?;

                wait(
                    self,
                    &DeviceState::Activated,
                    self.dbus_manager.method_timeout(),
                )
            }
        }
    }

    /// Disconnect a Network Manager device.
    pub fn disconnect(&self) -> Result<DeviceState> {
        let state = self.get_state()?;

        match state {
            DeviceState::Disconnected => Ok(DeviceState::Disconnected),
            _ => {
                self.dbus_manager.disconnect_device(&self.path)?;

                wait(
                    self,
                    &DeviceState::Disconnected,
                    self.dbus_manager.method_timeout(),
                )
            }
        }
    }
}

impl fmt::Debug for Device {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "Device {{ path: {:?}, interface: {:?}, device_type: {:?} }}",
            self.path, self.interface, self.device_type
        )
    }
}

pub trait PathGetter {
    fn path(&self) -> &str;
}

impl PathGetter for Device {
    fn path(&self) -> &str {
        &self.path
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum DeviceType {
    Unknown,
    Ethernet,
    WiFi,
    Unused1,
    Unused2,
    Bt,
    OlpcMesh,
    Wimax,
    Modem,
    Infiniband,
    Bond,
    Vlan,
    Adsl,
    Bridge,
    Generic,
    Team,
    Tun,
    IpTunnel,
    Macvlan,
    Vxlan,
    Veth,
    Macsec,
    Dummy,
    Ppp,
    OvsInterface,
    OvsPort,
    OvsBridge,
    Wpan,
    Lowpan,
    Wireguard,
    WifiP2p,
    Vrf,
}

impl From<i64> for DeviceType {
    fn from(device_type: i64) -> Self {
        match device_type {
            0 => DeviceType::Unknown,
            1 => DeviceType::Ethernet,
            2 => DeviceType::WiFi,
            3 => DeviceType::Unused1,
            4 => DeviceType::Unused2,
            5 => DeviceType::Bt,
            6 => DeviceType::OlpcMesh,
            7 => DeviceType::Wimax,
            8 => DeviceType::Modem,
            9 => DeviceType::Infiniband,
            10 => DeviceType::Bond,
            11 => DeviceType::Vlan,
            12 => DeviceType::Adsl,
            13 => DeviceType::Bridge,
            14 => DeviceType::Generic,
            15 => DeviceType::Team,
            16 => DeviceType::Tun,
            17 => DeviceType::IpTunnel,
            18 => DeviceType::Macvlan,
            19 => DeviceType::Vxlan,
            20 => DeviceType::Veth,
            21 => DeviceType::Macsec,
            22 => DeviceType::Dummy,
            23 => DeviceType::Ppp,
            24 => DeviceType::OvsInterface,
            25 => DeviceType::OvsPort,
            26 => DeviceType::OvsBridge,
            27 => DeviceType::Wpan,
            28 => DeviceType::Lowpan,
            29 => DeviceType::Wireguard,
            30 => DeviceType::WifiP2p,
            31 => DeviceType::Vrf,
            _ => {
                warn!("Undefined device type: {}", device_type);
                DeviceType::Unknown
            }
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum DeviceState {
    Unknown,
    Unmanaged,
    Unavailable,
    Disconnected,
    Prepare,
    Config,
    NeedAuth,
    IpConfig,
    IpCheck,
    Secondaries,
    Activated,
    Deactivating,
    Failed,
}

impl From<i64> for DeviceState {
    fn from(state: i64) -> Self {
        match state {
            0 => DeviceState::Unknown,
            10 => DeviceState::Unmanaged,
            20 => DeviceState::Unavailable,
            30 => DeviceState::Disconnected,
            40 => DeviceState::Prepare,
            50 => DeviceState::Config,
            60 => DeviceState::NeedAuth,
            70 => DeviceState::IpConfig,
            80 => DeviceState::IpCheck,
            90 => DeviceState::Secondaries,
            100 => DeviceState::Activated,
            110 => DeviceState::Deactivating,
            120 => DeviceState::Failed,
            _ => {
                warn!("Undefined device state: {}", state);
                DeviceState::Unknown
            }
        }
    }
}

pub fn get_devices(dbus_manager: &Rc<DBusNetworkManager>) -> Result<Vec<Device>> {
    let device_paths = dbus_manager.get_devices()?;

    let mut result = Vec::with_capacity(device_paths.len());

    for path in device_paths {
        let device = Device::init(dbus_manager, &path)?;

        result.push(device);
    }

    Ok(result)
}

pub fn get_device_by_interface(
    dbus_manager: &Rc<DBusNetworkManager>,
    interface: &str,
) -> Result<Device> {
    let path = dbus_manager.get_device_by_interface(interface)?;

    Device::init(dbus_manager, &path)
}

pub fn get_active_connection_devices(
    dbus_manager: &Rc<DBusNetworkManager>,
    active_path: &str,
) -> Result<Vec<Device>> {
    let device_paths = dbus_manager.get_active_connection_devices(active_path)?;

    let mut result = Vec::with_capacity(device_paths.len());

    for path in device_paths {
        let device = Device::init(dbus_manager, &path)?;

        result.push(device);
    }

    Ok(result)
}

fn wait(device: &Device, target_state: &DeviceState, timeout: u64) -> Result<DeviceState> {
    if timeout == 0 {
        return device.get_state();
    }

    debug!("Waiting for device state: {:?}", target_state);

    let mut total_time = 0;

    loop {
        ::std::thread::sleep(::std::time::Duration::from_secs(1));

        let state = device.get_state()?;

        total_time += 1;

        if state == *target_state {
            debug!(
                "Device target state reached: {:?} / {}s elapsed",
                state, total_time
            );

            return Ok(state);
        } else if total_time >= timeout {
            debug!(
                "Timeout reached in waiting for device state ({:?}): {:?} / {}s elapsed",
                target_state, state, total_time
            );

            return Ok(state);
        }

        debug!(
            "Still waiting for device state ({:?}): {:?} / {}s elapsed",
            target_state, state, total_time
        );
    }
}
